src/runtime/HalideRuntime.h

```diff
@@ -617,12 +619,32 @@ extern int halide_buffer_copy(void *user_context, struct halide_buffer_t *src,
  * care must be taken to update them together as needed. Note also
  * that device interfaces which support cropping may still not support
  * cropping a crop. Instead, create a new crop of the parent
- * buffer. */
+ * buffer. Note that src and dst are required to have the same number of
+ * dimensions. */
 extern int halide_device_crop(void *user_context,
                               const struct halide_buffer_t *src,
                               struct halide_buffer_t *dst);
 
-/** Release any resources associated with a cropped view of another
+/** Give the destination buffer a device allocation which is an alias
+ * for a similar coordinate range in the source buffer, but with one dimension
+ * sliced away in the dst. Modifies the device, device_interface, and the
+ * device_dirty flag only. Only supported by some device APIs (others will return
+ * halide_error_code_device_crop_unsupported). Call
+ * halide_device_release_crop instead of halide_device_free to clean
+ * up resources associated with the sliced view. Do not free the
+ * device allocation on the source buffer while the destination buffer
+ * still lives. Note that the two buffers do not share dirty flags, so
+ * care must be taken to update them together as needed. Note also
+ * that device interfaces which support cropping may still not support
+ * cropping a crop. Instead, create a new crop of the parent
+ * buffer. Note that the dst buffer must have exactly one fewer dimension than
```

Comment:
Given don't-take-a-crop-of-a-crop, how would one slice away multiple dimensions?

Reply:
From inspection of existing implementations it's not clear to me that this is the case for anything we actually support -- I've pinged @zvookin to see if there were real cases he anticipated. (If this is likely to be a real limitation somwehere then the slice API needs rethinking.)

Reply:
The idea is that if we end up supporting e.g. OpenGL subtextures, at the Halide::Runtime::Buffer level we can always crop from the root texture because the Buffer holds onto a reference to the original parent. I thought I had effectively implemented that, but apparently not. (Can't recall if I ran into an issue doing so or if I just decided not to do it because it wasn't necessary for any of the devices I was actually supporting crop for. I briefly looked at OpenGL texture stuff and like everything in OpenGL, it was going to be a massive hassle to get working.)

Unfortunately we can't do this for slicing because the top-level API requires slicing one dimension at a time. I guess we could remember the lists of slices applied to the parent with the reference to its buffer and then apply them all at once. That does require changing the device API to take an array of slices however. I'm not sure this is really worth addressing here.

Reply:
I'm going to just update the comments and move forward, as this is blocking some other issues.